src/Sarif.Viewer.VisualStudio/ErrorList/SarifTableDataSource.cs

```diff
@@ -109,91 +105,92 @@ public IDisposable Subscribe(ITableDataSink sink)
 
         public void AddSinkManager(SinkManager manager)
         {
-            // This call can, in theory, happen from any thread so be appropriately thread safe.
-            // In practice, it will probably be called only once from the UI thread (by the error list tool window).
-            lock (_managers)
+            using (this.sinkManagerLock.EnterWriteLock())
             {
-                _managers.Add(manager);
+                sinkManagers.Add(manager);
             }
         }
 
         public void RemoveSinkManager(SinkManager manager)
         {
-            // This call can, in theory, happen from any thread so be appropriately thread safe.
-            // In practice, it will probably be called only once from the UI thread (by the error list tool window).
-            lock (_managers)
+            using (this.sinkManagerLock.EnterWriteLock())
             {
-                _managers.Remove(manager);
+                sinkManagers.Remove(manager);
             }
         }
 
         public void UpdateAllSinks()
         {
-            lock (_managers)
+            this.CallSinkManagers((sinkManager) =>
             {
-                foreach (var manager in _managers)
+                IImmutableList<SarifResultTableEntry> entriesToNotify;
+
+                using (this.tableEntriesLock.EnterReadLock())
                 {
-                    manager.UpdateSink(_snapshots.Values);
+                    entriesToNotify = logFileToTableEntries.Values.SelectMany((tableEntries) => tableEntries).ToImmutableList();
                 }
-            }
+
+                sinkManager.AddEntries(entriesToNotify);
```

Comment:
This function needs to be deleted, and this code needs to happen inside the Subscribe method